### DISCUSSION TITLE
Feature request: mathematical morphology beyond binary images

### DISCUSSION THREAD

**Discussion by Leengit** [libvips/libvips#2345] [Q&A]
Thank you for an excellent package and, in particular, the [vips_morph](https://libvips.github.io/libvips/API/current/libvips-morphology.html#vips-morph) functionality that works on binary images (with pixel intensities of 0 and 255).  Additional functionality would be greatly appreciated!  That is, the morphology concepts are also applicable to grayscale images (and each channel of multi-chanel / color images); where the VIPS_OPERATION_MORPHOLOGY_ERODE functionality would implement a numerical MINimization instead of a logical AND, and the VIPS_OPERATION_MORPHOLOGY_DILATE functionality would implement a numerical MAXimization instead of a logical OR.  (Conveniently, MIN and MAX are 100% consistent with the current binary image functionality.)  You can read about these grayscale approaches all over the web; e.g., at [MathWorks: Types of Morphological Operations](https://www.mathworks.com/help/images/morphological-dilation-and-erosion.html) and at [Wikipedia: Mathematical Morphology -> Grayscale morphology](https://en.wikipedia.org/wiki/Mathematical_morphology#Grayscale_morphology).

Beyond these basic grayscale options, there are also addition morphological functions that would be useful.  The obvious first two are `opening`, which is erosion followed by dilation using the same mask, and `closing`, which is dilation followed by erosion using the same mask.  If you can implement these two too, that would be fabulous!

**Comment by jcupitt** [comment]
Hi @Leengit,

Thanks for your kind words. libvips has greyscale erode/dilate as `vips_rank()`:

https://libvips.github.io/libvips/API/current/libvips-morphology.html#vips-rank

Things like open and close and supposed to be done a level up, eg. in Python just:

```python
image = image.dilate(mask).erode(mask)
```

libvips has very fast function composition, so chaining operations like this is just as quick (almost) as a single custom operation would be.

**Reply by Leengit** [reply]
@jcupitt Thank you for the quick reply.  `vips_rank` does look quite close to what I was hoping for, except that it looks as though the mask that can be used with `vips_morph` is not available for `vips_rank`; for the latter I believe that the "mask" has to be a rectangle specified via `width` and `height`.

So, for some list of wishes for the future, I'd add having full mask functionality available with `vips_rank`.

**Reply by jcupitt** [reply]
That's true, there's no structuring element. Sure, let's tag this as an enhancement then.

**Comment by jcupitt** [comment]
enhancement issue: https://github.com/libvips/libvips/issues/2346
